File: backend/services/workflow-runner/pkg/notification/provider.go
/*
 * Copyright 2026 Leftbin/Stigmer
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package notification

import (
	"context"
	"fmt"
	"sync"
)

// NotificationRequest holds the resolved notification content.
type NotificationRequest struct {
	Channel    string
	Recipients []string
	Subject    string
	Body       string
	Template   string
	Metadata   map[string]string
}

// NotificationResult describes delivery outcome.
type NotificationResult struct {
	Channel     string   `json:"channel"`
	Recipients  []string `json:"recipients"`
	Delivered   bool     `json:"delivered"`
	DeliveredAt string   `json:"delivered_at,omitempty"`
	Error       string   `json:"error,omitempty"`
}

// Provider sends notifications through a specific channel.
type Provider interface {
	Send(ctx context.Context, req NotificationRequest) (*NotificationResult, error)
	Channel() string
}

var (
	mu        sync.RWMutex
	providers = map[string]Provider{}
)

// Register adds a provider for the given channel.
func Register(p Provider) {
	mu.Lock()
	defer mu.Unlock()
	providers[p.Channel()] = p
}

// Get returns the provider for the given channel, or an error if not found.
func Get(channel string) (Provider, error) {
	mu.RLock()
	defer mu.RUnlock()
	p, ok := providers[channel]
	if !ok {
		return nil, fmt.Errorf("notification channel '%s' is not implemented; available channels: webhook", channel)
	}
	return p, nil
}

File: backend/services/runner/src/notification/provider.ts
/**
 * Notification provider registry.
 *
 * Channels are string identifiers (not an enum) for extensibility —
 * new channels can be added without code changes beyond implementing
 * the Provider interface and calling registerProvider().
 */

export interface NotificationRequest {
  readonly channel: string;
  readonly recipients: readonly string[];
  readonly subject: string;
  readonly body: string;
  readonly template?: string;
  readonly metadata: Readonly<Record<string, string>>;
}

export interface NotificationResult {
  readonly channel: string;
  readonly recipients: readonly string[];
  readonly delivered: boolean;
  readonly delivered_at?: string;
  readonly error?: string;
}

export interface NotificationProvider {
  channel(): string;
  send(request: NotificationRequest): Promise<NotificationResult>;
}

const providers = new Map<string, NotificationProvider>();

export function registerProvider(provider: NotificationProvider): void {
  providers.set(provider.channel(), provider);
}

export function getProvider(channel: string): NotificationProvider {
  const provider = providers.get(channel);
  if (!provider) {
    const available = providers.size > 0
      ? Array.from(providers.keys()).join(", ")
      : "(none)";
    throw new Error(
      `Notification channel '${channel}' is not implemented; available channels: ${available}`,
    );
  }
  return provider;
}

export function resetProviders(): void {
  providers.clear();
}
